### DISCUSSION TITLE
2011/06/02/modkit-%E4%B8%80%E7%A7%8D%E5%9B%BE%E5%BD%A2%E5%8C%96%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/blog#331] [Announcements]
# 2011/06/02/modkit-%E4%B8%80%E7%A7%8D%E5%9B%BE%E5%BD%A2%E5%8C%96%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80

目前，像LabVIEW这样功能强大，比较通用的图形化编程语言可能绝无仅有的。但市面上，比较小型的，针对特定应用的图形化编程语言却五花八门，随处可见。今天同事又给我介绍了一款，叫做Modkit，官方网站：http://www.modk.it （河蟹们访问需翻墙）。

https://blog.qizhen.xyz/2011/06/02/modkit-%E4%B8%80%E7%A7%8D%E5%9B%BE%E5%BD%A2%E5%8C%96%E7%BC%96%E7%A8%8B%E8%AF%AD%E8%A8%80

**Comment by ruanqizhen** [comment]
之前博客下的留言：

习惯孤独
2011年06月4日 14:42 编辑
您好：
我想问您一个labview的问题，我在调用函数动态链接库的时候出现“调用库函数节点存在错误 该错误可能引起内存损坏 ”我不知道问题出在哪儿..下年是底层的VC代码：
LRESULT CANReadFile
(
WORDPort,
DWORD NumNeeded,
PVOID FAR RcvBuf,
PDWORD pNumReturned
)

Purpose
Read the received data from receive buffer. This function allow user receive data by interrupt way.
Parameters

Name Direction Type Range Description
Port Input WORD default Specify the port number
NumNeeded Input DWORD default The message number that should be read from buffer
RcvBuf OutPut PVOID FAR default The receive buffer that the transmit data stored
pNumReturned Output PDWORD default If pNumReturned = 0, the function is called unsuccessful. Else its value is the received message number.

Return
SUCCESS if successful
CANInvalidPortIndex ifport index is invalid, must from 1 to 10
CANPortIsNotOpened ifport is not opened

Example:

#include “canbus2.0.h”

typedef struct{
UCHAR rtr;
ULONG id;
UCHAR dlen;
UCHAR data[8];
} CAN_MSG_T;

main( )
{
CAN_MSG_T msg;
DWORD cbReturned = 0;
int nLength;
WORD m_wProtocolType;
UINT m_Port;
string m_RxBuf;

// do some initiate works
……

// read the data from receive buffer
if ( m_wProtocolType = = CANBUS_PROTOCOL_20A )
{
CANReadFile( m_Port, 1, &msg, &cbReturned );
if ( cbReturned )
{
m_RxID = msg.id;

// get receive data length
nLength = msg.dlen;
if ( msg.dlen!= 0 )
{
printf( “%s”, (PUCHAR)msg.data );
}
}
}
}
我在编程是将pvoid far参数格式设置成了按值处理 连接的是一个簇 就出现了上述的问题，请您这样的高手帮帮忙~

回复
ruanqizhen
2011年06月9日 11:46 编辑
“调用库函数节点存在错误 该错误可能引起内存损坏 ” 这种错误一般是CLN节点没有设置正确引起的。
如果程序的C代码是你自己写的，我建议你把struct数据类型拆成简单数据类型与LabVIEW交互。简单类型相对来书不容易出错。

回复
